{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curated dataset ALS for LD SNPs predicition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-21 16:27:30.468881: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2025-02-21 16:27:30.468944: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2025-02-21 16:27:30.652544: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2025-02-21 16:27:31.088758: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-02-21 16:27:37.719042: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "# Hugging Face Datasets library\n",
    "from datasets import load_dataset, Dataset\n",
    "\n",
    "# Scikit-learn for evaluation and data splitting\n",
    "from sklearn import metrics\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Biopython for sequence processing\n",
    "from Bio import SeqIO\n",
    "\n",
    "# PyBedTools for genomic interval manipulation\n",
    "from pybedtools import BedTool\n",
    "\n",
    "# Data analysis and scientific computation\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "\n",
    "# PyTorch for deep learning\n",
    "import torch\n",
    "from torch import nn\n",
    "\n",
    "# Visualization libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Utilities and external libraries\n",
    "import os\n",
    "import pickle\n",
    "import genopyc\n",
    "import utility_functions as uf # Custom utility functions for loading models/tokenizers, datasets from the experiments\n",
    "\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load datasete raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variation ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Distance from rs10122902</th>\n",
       "      <th>Consequence Type</th>\n",
       "      <th>r2</th>\n",
       "      <th>D'</th>\n",
       "      <th>Phenotypes</th>\n",
       "      <th>GWAS_Trait_Category</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs10122902</td>\n",
       "      <td>9:27556782-27556782</td>\n",
       "      <td>0</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>undefined</td>\n",
       "      <td>undefined</td>\n",
       "      <td>Amyotrophic lateral sclerosis;Amyotrophic late...</td>\n",
       "      <td>Other trait</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs72727506</td>\n",
       "      <td>9:27541834-27541834</td>\n",
       "      <td>-14948</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No trait reported</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4879564</td>\n",
       "      <td>9:27557835-27557835</td>\n",
       "      <td>1053</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.956790</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No trait reported</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4878487</td>\n",
       "      <td>9:27552489-27552489</td>\n",
       "      <td>-4293</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.956790</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No trait reported</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs36038669</td>\n",
       "      <td>9:27536735-27536735</td>\n",
       "      <td>-20047</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>0.844444</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No trait reported</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs67483205</td>\n",
       "      <td>9:27541914-27541914</td>\n",
       "      <td>-14868</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No trait reported</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs3739526</td>\n",
       "      <td>9:27547315-27547315</td>\n",
       "      <td>-9467</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>FRONTOTEMPORAL DEMENTIA AND/OR AMYOTROPHIC LAT...</td>\n",
       "      <td>Neurological disorder</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Variation ID             Location  Distance from rs10122902  \\\n",
       "0   rs10122902  9:27556782-27556782                         0   \n",
       "1   rs72727506  9:27541834-27541834                    -14948   \n",
       "2    rs4879564  9:27557835-27557835                      1053   \n",
       "3    rs4878487  9:27552489-27552489                     -4293   \n",
       "4   rs36038669  9:27536735-27536735                    -20047   \n",
       "5   rs67483205  9:27541914-27541914                    -14868   \n",
       "6    rs3739526  9:27547315-27547315                     -9467   \n",
       "\n",
       "      Consequence Type         r2         D'  \\\n",
       "0   synonymous_variant  undefined  undefined   \n",
       "1   intergenic_variant   0.916667   1.000000   \n",
       "2       intron_variant   0.956790   1.000000   \n",
       "3       intron_variant   0.956790   1.000000   \n",
       "4   intergenic_variant   0.844444   1.000000   \n",
       "5   intergenic_variant   0.916667   1.000000   \n",
       "6  3_prime_UTR_variant   1.000000   1.000000   \n",
       "\n",
       "                                          Phenotypes    GWAS_Trait_Category  \\\n",
       "0  Amyotrophic lateral sclerosis;Amyotrophic late...            Other trait   \n",
       "1                                                NaN      No trait reported   \n",
       "2                                                NaN      No trait reported   \n",
       "3                                                NaN      No trait reported   \n",
       "4                                                NaN      No trait reported   \n",
       "5                                                NaN      No trait reported   \n",
       "6  FRONTOTEMPORAL DEMENTIA AND/OR AMYOTROPHIC LAT...  Neurological disorder   \n",
       "\n",
       "    Source  \n",
       "0  Ensembl  \n",
       "1  Ensembl  \n",
       "2  Ensembl  \n",
       "3  Ensembl  \n",
       "4  Ensembl  \n",
       "5  Ensembl  \n",
       "6  Ensembl  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all file in a directory\n",
    "path_datasets = \"/data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/\"\n",
    "files = os.listdir(path_datasets)\n",
    "files = [file for file in files if file.endswith('.tsv')]\n",
    "\n",
    "## read tsv file\n",
    "path_dataset = path_datasets+files[0]\n",
    "df = pd.read_csv(path_dataset, sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get SNPs info: chromosome, position, reference, alternative, snp_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Variation ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27541914</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>rs67483205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27557835</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs4879564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27536735</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs36038669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27541834</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs72727506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27552489</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs4878487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27547315</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs3739526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>27556782</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs10122902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Position Reference Alternative Variation ID\n",
       "0       chr9  27541914         C           A   rs67483205\n",
       "1       chr9  27557835         C           T    rs4879564\n",
       "2       chr9  27536735         A           G   rs36038669\n",
       "3       chr9  27541834         C           T   rs72727506\n",
       "4       chr9  27552489         T           C    rs4878487\n",
       "5       chr9  27547315         G           C    rs3739526\n",
       "6       chr9  27556782         G           A   rs10122902"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def process_snp_candidate(list_snps):\n",
    "    local_data = []\n",
    "    try:\n",
    "        # Fetch variant information\n",
    "        info_ld = genopyc.get_variants_info(list_snps, chunked=False, chunksize=200)\n",
    "        \n",
    "        # Extract and filter relevant data\n",
    "        for snp_id, snp_info in info_ld.items():\n",
    "            mappings = snp_info.get('mappings', [])\n",
    "            if not mappings:\n",
    "                continue\n",
    "\n",
    "            allele_string = mappings[0].get('allele_string', '')\n",
    "            location = mappings[0].get('location', '')\n",
    "            if not allele_string or not location:\n",
    "                continue\n",
    "\n",
    "            # Parse alleles and position\n",
    "            alleles = allele_string.split('/')\n",
    "            pos = location.split(':')\n",
    "            #if len(pos) != 2 or len(alleles) != 2:\n",
    "            #    continue\n",
    "\n",
    "            chromosome = f\"chr{pos[0]}\"\n",
    "            position = int(pos[1].split('-')[0])\n",
    "            reference = alleles[0]\n",
    "\n",
    "            alternative = snp_info.get('minor_allele', None)\n",
    "            if alternative is None:\n",
    "                alternative = alleles[1]\n",
    "                \n",
    "            if reference == '-' or alternative=='-':\n",
    "                continue\n",
    "\n",
    "            # Include only SNPs (single nucleotide changes)\n",
    "            #if len(reference) == 1 and len(alternative) == 1:\n",
    "            local_data.append([chromosome, position, reference, alternative, snp_id])\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {list_snps}: {e}\")\n",
    "    return local_data\n",
    "\n",
    "\n",
    "## create a df with the SNPs\n",
    "df_snps = pd.DataFrame(process_snp_candidate(df['Variation ID'].tolist()), columns=['Chromosome', 'Position', 'Reference', 'Alternative', 'Variation ID'])\n",
    "\n",
    "df_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main: make it for every dataset in the folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs10122902.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs1031153.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs117761967.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs2453555.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs2814707.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs3849942.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs3849943.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "total number of chunks: 2\n",
      "chunk 0 processed\n",
      "chunk 1 processed\n",
      "rs73440933.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n",
      "rs774359.tsv saved in /data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/curated/\n"
     ]
    }
   ],
   "source": [
    "## get all file in a directory\n",
    "path_datasets = \"/data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/\"\n",
    "files = os.listdir(path_datasets)\n",
    "files = [file for file in files if file.endswith('.tsv')]\n",
    "\n",
    "## create a folder curated if not exists\n",
    "path_to_save = path_datasets+'curated/'\n",
    "if not os.path.exists(path_datasets+'curated/'):\n",
    "    os.makedirs(path_to_save)\n",
    "\n",
    "## save the df\n",
    "for file in files: \n",
    "    path_dataset = path_datasets+file\n",
    "    df = pd.read_csv(path_dataset, sep='\\t')\n",
    "    df_snps = pd.DataFrame(process_snp_candidate(df['Variation ID'].tolist()), columns=['Chromosome', 'Position', 'Reference', 'Alternative', 'Variation ID'])\n",
    "    df_snps.to_csv(path_to_save+file, sep='\\t', index=False)\n",
    "    print(f\"{file} saved in {path_to_save}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'predictions_alt': array([[3.9635066e-02],\n",
       "        [3.5207569e-05],\n",
       "        [2.1966679e-03],\n",
       "        [1.4339499e-03],\n",
       "        [6.6309660e-03],\n",
       "        [6.4007536e-04],\n",
       "        [3.6650585e-05]], dtype=float32),\n",
       " 'predictions_ref': array([[2.0876354e-02],\n",
       "        [3.4290199e-05],\n",
       "        [2.3022976e-03],\n",
       "        [9.1243640e-04],\n",
       "        [2.2561645e-02],\n",
       "        [6.6661910e-04],\n",
       "        [4.7790218e-05]], dtype=float32)}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## read pickle file \n",
    "with open('/data/Dcode/gaetano/repos/AI4Genomic/data/ALS/gene_c9/LDs/predictions/trednet_LD_C9_ALS_rs1031153.tsv.pkl', 'rb') as f:\n",
    "    df = pickle.load(f, encoding='latin1')\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
